### ISSUE TITLE
Human-Shaped Code

### ISSUE THREAD

**Issue by rosemarydotworld** [SassConf/2015-austin-speaker-cfp#4]
# Human-Shaped Code
## Type of Presentation

`[x]` Standard Length Talk
`[ ]` Lightning Talk
`[ ]` Workshop
`[ ]` Moderated Discussion
## Description

Let's explore code through the lens of interface design. How does code compare with other interfaces? Where does misuse and misunderstanding come from? How can we use what we know about memory, attention, and language comprehension to build a more harmonious relationship with code?

I love code, despite the harm I know it can do to me. I've wracked my brain over opaque, undocumented libraries. I've spent sleepless nights trying in vain to fix bugs in overgrown codebases. I've shed sweat, blood, and tears over code. I do it because it's my job. But I think it can be better.

Code is a tool. If one of our tools hurts to use—if the handle of our hammer gives us a blister—we need a better designed tool. The problems we have with code stem from a lack of understanding. The same way that our crappy hammer doesn't fit our hand, the way we write code doesn't anticipate the shape of our minds. With a better model of how we think about code, we can approach our work in a person-first way and facilitate understanding, both in ourselves and the people around us.

We can craft codebases that are intuitive and friendly; that are approachable to rookies and accommodating to veterans. We can save ourselves and our colleagues time and energy. We can ensure that our good work will not go to waste. We just need to design human-shaped code.

Also included: animated GIFs, Buckminster Fuller, and unbridled enthusiasm.

---

_I've been doing internal talks about code quality for years (some of which [Hampton](http://twitter.com/hcatlin) has seen), but this is a much broader topic than I usually take on. I've got a stack of research materials about a foot high, and I'm workshopping with my bosses, [Mike](https://twitter.com/monteiro) and [Erika](https://twitter.com/mulegirl), to make the talk as incisive and practical as it can be._

_Basically: I'm really excited for you to see it._
## Speaker Info
- Name      : Liam Campbell
- Location  : San Francisco, CA
- Contact   : liam@muledesign.com
##### Social Media:
- Twitter   : [@liamdanger](http://twitter.com/liamdanger)
- GitHub    : [liamdanger](http://github.com/liamdanger)
- Url(s)    : [**Commentary**, my code blog](http://commentary.link)
##### Bio:

I'm a web developer, designer, and enthusiast currently working at Mule Design as their lead developer. Let's learn together.
##### Photo:

![Avatar](https://avatars1.githubusercontent.com/u/1302256?v=3&s=460)

**Comment by scottkellum** [comment]
Hey, really liking this. Curious what your thoughts are on documentation. Making code that is self-teaching is a problem I have run into and while GUIs allow people to mess around and learn more easily, building knowledge on top of code has proven to be difficult. Using the tool metaphor, a hammer has a handle, it’s usually clear where you’re souped to hold it without a readme.

**Comment by rosemarydotworld** [comment]
Good question! Self-documenting code is a useful ideal to aim for, but I don't think that any system over a certain level of complexity can ever be completely self-documenting. There are too many variables about how people think, what their experience leads them to believe, and even what their culture tells them about a tool. Making a universally intuitive tool is impossible—I'd be willing to bet that some people out there really _do_ need a readme for their hammer.

That said, I consider labelling and commenting a form of inline documentation as well as a part of an interface's design. Any GUI with more than two buttons has _some_ system of labelling, and there's a lot of concerns that go into designing one of those systems. How much does the user know? How much labelling is already in this area to distract or obfuscate this meaning<sup>**1**</sup>? What's the clearest way to phrase an instruction? These are the same problems we encounter when commenting code.

The prevailing wisdom is that the better designed your inline documentation is, the less you need that readme, and the more willing users are to experiment with the tools provided to them. In the absence of a first-hand source of knowledge, the tool itself needs to be approachable and comprehensible enough that it facilitates the kind of exploration that leads to deep understanding.

Hope that helps!

---

<sup>**1**</sup> The more documentation you have, the less useful _any_ of it can be. It becomes cognitively difficult to sort through, and people become more likely to dismiss it out of hand, even if it happens to contain the exact information they need.

**Comment by misscs** [comment]
Very compelling proposal. As an attendee, I would be jazzed to see this. As a curator, I would like to know more about what an audience member can expect to walk away with. What topics or ideas are you going to dive into? What is the "better model" you mention—what things comprise it?

**Comment by rosemarydotworld** [comment]
That's a good question, the answer to which is still not fully baked. I'll give you as much as I can, and I hope you can forgive my early-stage vagueness.

The end goal I have for the talk is to produce a set of design principles that people can use to evaluate and refine their code—kind of like code smells for usability. I'm not expecting any of them to be revolutionary, or even particularly novel for someone who already writes well-formed code. Part of the job I'm trying to do here is give ammunition to the things we hold up as coding best practices.

I'm getting that ammunition by researching behavioral psychology, cognitive psychology, linguistics, and interface design. 

**Here's a for-instance:** we mostly agree that [short methods](https://github.com/bbatsov/ruby-style-guide#short-methods) and [chunking different concerns](https://github.com/bbatsov/ruby-style-guide#empty-lines-between-methods) are good ideas, but we can back those ideas up with research on cognitive load, and how our rational thinking skills are reduced when we have to keep too much in our head at once. The research can then help us deepen our knowledge of why these best practices work, and come up with a more nuanced approach to organizing code.

And that's really why I'm not writing this as a straightforward set of heuristics about code readability. I think there's much more interesting work to be done if we approach the design of code as a legitimate subfield of design. There's a lot to learn and discover, and I hope that by giving people the "thought technology" to look at their code through this lens, they can come up with way more interesting and useful insights than I'd ever be able to muster.

---

I hope that clarifies rather than muddles the topic, or at the very least, muddles it in an interesting way. The structure of this talk is going to be heavily affected by what I can learn in research, and if it's selected, I'd really look forward to the help the mentoring team can provide towards making the talk practical as well as theoretically interesting.

**Comment by elyseholladay** [comment]
Hi Liam!

Thank you so much for submitting to SassConf this year!

Unfortunately, we weren’t able to select your talk. 

We had an incredible number of submissions this year: 81, in fact, enough to fill up over two weeks of Sassy goodness! But we only have two days, and we couldn’t pick everything.

If you have any questions at all about our selection process, your submission, or anything else at all, please reach out: elyse@sassconf.com and I’ll gladly give you more details.

Again, thank you for submitting. It’s people like you, who are willing to put themselves out there and work hard and submit and give talks that make it possible to even have SassConf. I hope you will submit again next year and continue to be part of the Sass community!

See you in November!
